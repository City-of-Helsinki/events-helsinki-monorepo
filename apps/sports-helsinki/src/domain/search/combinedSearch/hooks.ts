import { useLazyQuery } from '@apollo/client';
import type {
  SearchListQuery,
  EventListQuery,
} from 'events-helsinki-components';
import {
  SearchListDocument,
  EventTypeId,
  EventListDocument,
} from 'events-helsinki-components';
import { useRouter } from 'next-router-mock';
import React, { useRef, useEffect } from 'react';
import AppConfig from 'domain/app/AppConfig';
import useUnifiedSearchVariables from 'domain/unifiedSearch/useUnifiedSearchVariables';
import { useEventSearchFilters } from '../eventSearch/SearchPage';
import type { SearchTabId } from './searchTabs/tabsContext';
import { isSearchTabId, useTabsContext } from './searchTabs/tabsContext';

/**
 * Use an URL parameter to scroll to
 * the previously used search result card on a page change.
 */
export const useScrollToSearchResultItem = () => {
  const router = useRouter();
  const scrollTo = router.query?.scrollTo;
  const listRef = useRef<HTMLUListElement | null>(null);

  useEffect(() => {
    const listElement = listRef.current;

    if (scrollTo) {
      const listItemElement = listElement?.querySelector(
        decodeURIComponent(scrollTo.toString())
      );

      if (listItemElement) {
        listItemElement.scrollIntoView({
          block: 'center',
        });
      }
    }
  }, [scrollTo]);
};

export const useSearchTabsWithParams = (defaultTab: SearchTabId) => {
  const router = useRouter();
  const searchParams = new URLSearchParams(router.asPath.split('?')[1]);
  const searchTypeParam = searchParams.get('searchType');
  const initTab = searchTypeParam
    ? (searchTypeParam as SearchTabId)
    : defaultTab;

  // If the search type param is not given in the URL, set it there.
  React.useEffect(() => {
    if (!isSearchTabId(searchTypeParam)) {
      router.push(
        { query: { ...router.query, searchType: defaultTab } },
        undefined,
        { shallow: true }
      );
    }
  }, [searchTypeParam, router, defaultTab]);

  return { initTab, searchTypeParam };
};

export function useLazyVenueSearchForTabCount() {
  const { setResultCount } = useTabsContext();
  const venueSearchFilters = useUnifiedSearchVariables();
  const variables = React.useMemo(
    () => ({
      ...venueSearchFilters,
      includeHaukiFields: AppConfig.isHaukiEnabled,
    }),
    [venueSearchFilters]
  );
  const [search, { loading, data, ...delegatedProps }] = useLazyQuery(
    SearchListDocument,
    {
      variables,
      // FIXME: Set the fetch policy to not trigger cache,
      // or the Apollo client fails to request
      // same query multiple times. Depending on settings,
      // the client stops the query when the first response is given,
      // and uses the cache for the next query with same name or
      // returns the items twice.
      // Since the query is exactly the same as the actual listing query,
      // it would be really nice to have the result in cache already.
      // See https://community.apollographql.com/t/executing-the-same-query-multiple-times-with-different-variables/3052
      fetchPolicy: 'no-cache',
      ssr: false,
    }
  );

  React.useEffect(() => {
    const count = () => {
      if (loading || !data) {
        return null;
      }
      return (data as SearchListQuery)?.unifiedSearch?.count ?? 0;
    };
    setResultCount('Venue', count());
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [data, loading]);

  return { search, loading, data, ...delegatedProps };
}

export function useLazyEventSearchForTabCount({
  eventType,
}: {
  eventType: EventTypeId;
}) {
  const { setResultCount } = useTabsContext();
  const variables = useEventSearchFilters(eventType);
  const [search, { loading, data, ...delegatedProps }] = useLazyQuery(
    EventListDocument,
    {
      variables,
      // FIXME: Set the fetch policy to not trigger cache,
      // or the Apollo client fails to request
      // same query multiple times. Depending on settings,
      // the client stops the query when the first response is given,
      // and uses the cache for the next query with same name or
      // returns the items twice.
      // Since the query is exactly the same as the actual listing query,
      // it would be really nice to have the result in cache already.
      // See https://community.apollographql.com/t/executing-the-same-query-multiple-times-with-different-variables/3052
      fetchPolicy: 'no-cache',
      ssr: false,
    }
  );

  React.useEffect(() => {
    const count = () => {
      if (loading || !data) {
        return null;
      }
      return ((data as EventListQuery)?.eventList?.meta.count as number) ?? 0;
    };
    setResultCount(eventType, count());
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [data, loading, eventType]);

  return { search, loading, data, ...delegatedProps };
}

export function useSearchTabResultCounts() {
  const { search: venueSearch } = useLazyVenueSearchForTabCount();
  const { search: generalEventSearch } = useLazyEventSearchForTabCount({
    eventType: EventTypeId.General,
  });
  const { search: courseEventSearch } = useLazyEventSearchForTabCount({
    eventType: EventTypeId.Course,
  });
  React.useEffect(() => {
    venueSearch();
    generalEventSearch();
    courseEventSearch();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);
}
