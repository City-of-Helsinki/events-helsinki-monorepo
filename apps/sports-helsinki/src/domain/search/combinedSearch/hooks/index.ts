export { default as useLazyEventSearchForTabCount } from './useLazyEventSearchForTabCount';
export { default as useLazyVenueSearchForTabCount } from './useLazyVenueSearchForTabCount';
export { default as useScrollToSearchResultItem } from './useScrollToSearchResultItem';
export { default as useSearchTabResultCounts } from './useSearchTabResultCounts';
export { default as useSearchTabsWithParams } from './useSearchTabsWithParams';
