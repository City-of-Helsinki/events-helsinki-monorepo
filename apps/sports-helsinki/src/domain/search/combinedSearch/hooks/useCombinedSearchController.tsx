import useGeolocation from '@events-helsinki/components/geolocation/useGeolocation';
import useLocale from '@events-helsinki/components/hooks/useLocale';
import { useRouter } from 'next/router';
import type { CombinedSearchContextType } from '../adapters/CombinedSearchContext';
import CombinedSearchFormAdapter from '../adapters/CombinedSearchFormAdapter';

export function useCombinedSearchController() {
  const locale = useLocale();
  const router = useRouter();
  const geolocation = useGeolocation({ skip: true });
  // FIXME: use import { useSearchParams } from 'next/navigation' if it is fixed so that it is initialized as soon as the asPath is.
  const searchParams = new URLSearchParams(router.asPath.split('?')[1]);
  const combinedSearchFormAdapter = new CombinedSearchFormAdapter(
    locale,
    searchParams,
    geolocation.coordinates
  );
  const updateRouteToSearchPage = (
    options?: Parameters<
      CombinedSearchContextType['updateRouteToSearchPage']
    >[0]
  ) => {
    const query = combinedSearchFormAdapter.getURLQuery();
    (async () => await router.push({ query }, undefined, options))();
  };
  return {
    combinedSearchFormAdapter,
    updateRouteToSearchPage,
  };
}
