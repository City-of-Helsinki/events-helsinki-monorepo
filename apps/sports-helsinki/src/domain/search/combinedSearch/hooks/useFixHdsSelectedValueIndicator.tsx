import React from 'react';

/**
 * The HDS-select has an issue with value indicator and the clearable-feature:
 * If the select-component value is set from external state and the form value is cleared,
 * the value indicator is not refreshed. The indicator tells that the value is still given,
 * when it's actualyl not.
 *
 * This hook fixes the issue by finding the clear-button and triggering it's onclick handler by clicking it.
 * @param selectComponentId the id-property value of the HDS select component.
 * @param value the current selected value (in state context).
 */
export default function useFixHdsSelectedValueIndicator({
  selectComponentId,
  value,
}: {
  selectComponentId: string;
  value: string | null | undefined;
}) {
  // Sync the value indicator when the select component value is cleared externally.
  React.useEffect(() => {
    // The select-component's most critical rendered element is the selector-button...
    const accessibilityShortcomingSelectorButton =
      window.document.getElementById(`${selectComponentId}-toggle-button`);
    // If a value is given and the clearable-feature is on,
    // there should be another button, that is a sibling element, to clear the value.
    const buttons =
      accessibilityShortcomingSelectorButton?.parentElement?.getElementsByTagName(
        'button'
      );
    const selectedValueIndicator =
      buttons?.length === 2 ? buttons[1] : undefined;

    // If the formValue is null or undefined, the indicator should not be visible.
    if (selectedValueIndicator && !value) {
      // Currently it seems like the only way to clear
      // the indicator is to trigger it's onclick-handler by clicking it.
      selectedValueIndicator.click();
    }
  }, [value, selectComponentId]);
}
