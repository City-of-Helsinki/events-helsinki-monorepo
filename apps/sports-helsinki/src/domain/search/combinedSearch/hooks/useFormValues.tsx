import { useSearchTranslation } from '@events-helsinki/components';
import React from 'react';
import {
  getTargetGroupOptions,
  getSportsCategoryOptions,
  sortExtendedCategoryOptions,
} from '../../eventSearch/utils';
import { useCombinedSearchContext } from '../adapters/CombinedSearchContext';

export function useFormValues() {
  const { t } = useSearchTranslation();
  const { formValues } = useCombinedSearchContext();
  const [autosuggestInput, setAutosuggestInput] = React.useState(
    formValues.text ?? ''
  );
  const [selectedSportsCategories, setSelectedSportsCategories] =
    React.useState<string[]>(formValues.sportsCategories);
  const [sportsCategoryInput, setSportsCategoryInput] = React.useState('');
  const sportsCategories = getSportsCategoryOptions(t).sort(
    sortExtendedCategoryOptions
  );
  const [selectedTargetGroups, setSelectedTargetGroups] = React.useState<
    string[]
  >(formValues.targetGroups);
  const [targetGroupInput, setTargetGroupInput] = React.useState('');
  const targetGroups = getTargetGroupOptions(t);

  // On page load, initialize the form with values
  // that are available only after the page has fully loaded
  React.useEffect(() => {
    setAutosuggestInput(formValues.text ?? '');
    setSelectedSportsCategories(formValues.sportsCategories);
    setSelectedTargetGroups(formValues.targetGroups);
  }, [formValues.sportsCategories, formValues.targetGroups, formValues.text]);

  return {
    autosuggestInput,
    setAutosuggestInput,
    selectedSportsCategories,
    setSelectedSportsCategories,
    sportsCategoryInput,
    setSportsCategoryInput,
    sportsCategories,
    selectedTargetGroups,
    setSelectedTargetGroups,
    targetGroupInput,
    setTargetGroupInput,
    targetGroups,
  };
}
