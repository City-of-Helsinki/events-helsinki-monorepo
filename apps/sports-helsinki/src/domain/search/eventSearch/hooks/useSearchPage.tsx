import {
  EventTypeId,
  useSearchTranslation,
  useIsSmallScreen,
  useEventListQuery,
  getLargeEventCardId,
  useClearClosedEventsFromApolloCache,
} from '@events-helsinki/components';
import { useRouter } from 'next/router';
import React from 'react';
import { useConfig } from 'react-helsinki-headless-cms';
import { scroller } from 'react-scroll';
import { toast } from 'react-toastify';
import { useCombinedSearchContext } from 'domain/search/combinedSearch/adapters/CombinedSearchContext';
import { SEARCH_ROUTES } from '../../../../constants';
import routerHelper from '../../../../domain/app/routerHelper';
import type { SearchPage } from '../../../../domain/search/combinedSearch/types';
import { getNextPage } from '../utils';

function useEventSearchPageQuery(eventType: EventTypeId) {
  const { searchVariables } = useCombinedSearchContext();
  const { t } = useSearchTranslation();
  const [isFetchingMore, setIsFetchingMore] = React.useState(false);

  // Query for the primary search / active search tab
  const { data, fetchMore, ...query } = useEventListQuery({
    ssr: false,
    variables:
      eventType === EventTypeId.Course
        ? searchVariables.course
        : searchVariables.event,
  });

  const handleLoadMore = async () => {
    const page = data?.eventList.meta ? getNextPage(data.eventList.meta) : null;
    setIsFetchingMore(true);
    if (page) {
      try {
        await fetchMore({
          variables: {
            page,
          },
        });
      } catch (e) {
        toast.error(t('search:errorLoadMore'));
      }
    }
    setIsFetchingMore(false);
  };

  // Clear the cache from the events of the past
  useClearClosedEventsFromApolloCache(data);

  return {
    data,
    fetchMore,
    ...query,
    isFetchingMore,
    handleLoadMore,
  };
}

function useSearchPage({ eventType }: { eventType: EventTypeId }): SearchPage {
  const router = useRouter();

  const isSmallScreen = useIsSmallScreen();
  const { meta } = useConfig();
  const {
    data: eventsData,
    loading: isLoadingEvents,
    isFetchingMore,
    handleLoadMore,
  } = useEventSearchPageQuery(eventType);

  const scrollToResultList = () => {
    if (isSmallScreen) {
      scroller.scrollTo('resultList', {
        delay: 0,
        duration: 1000,
        offset: -50,
        smooth: true,
      });
    }
  };

  const scrollToResultCard = (id: string) => {
    scroller.scrollTo(id, {
      delay: 0,
      duration: 300,
      offset: -50,
      smooth: true,
    });
  };

  const initialPageOnLoad = () => {
    if (router.asPath && router.query?.scrollToResults) {
      scrollToResultList();
    } else if (router.query?.eventId) {
      scrollToResultCard(
        getLargeEventCardId(
          Array.isArray(router.query.eventId)
            ? router.query.eventId[0]
            : router.query.eventId
        )
      );
      routerHelper.removeQueryParamsFromRouter(
        router,
        ['eventId'],
        SEARCH_ROUTES.SEARCH
      );
    }
  };

  const count = eventsData?.eventList?.meta.count ?? 0;
  const hasNext = !!eventsData?.eventList?.meta.next;

  return {
    meta,
    isSmallScreen,
    handleLoadMore,
    isFetchingMore,
    isLoading: isLoadingEvents,
    scrollToResultList,
    scrollToResultCard,
    resultList: eventsData?.eventList,
    initialPageOnLoad,
    count,
    hasNext,
  };
}

export default useSearchPage;
