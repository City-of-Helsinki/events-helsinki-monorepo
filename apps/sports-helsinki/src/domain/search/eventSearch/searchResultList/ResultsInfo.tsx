import type { EventTypeId } from '@events-helsinki/components';
import { IconSearch } from 'hds-react';

import { useTranslation } from 'next-i18next';
import React from 'react';

import styles from './resultsInfo.module.scss';

const ResultsInfoContainer: React.FC<{
  resultsCount: number;
  itemType?: EventTypeId | 'Venue';
}> = ({ resultsCount }) => {
  const { t } = useTranslation('search');

  if (resultsCount === 0) {
    return <ResultsInfo bigText={t(`searchNotification.noResultsTitle`)} />;
  }

  if (resultsCount < 5) {
    return <ResultsInfo bigText={t(`searchNotification.fewResultsTitle`)} />;
  }

  return null;
};

const ResultsInfo: React.FC<{
  smallText?: string;
  bigText: string;
}> = ({ bigText, smallText }) => {
  return (
    <div className={styles.noResultsInfo}>
      <div className={styles.iconWrapper}>
        <IconSearch aria-hidden />
      </div>
      <div className={styles.bigText}>{bigText}</div>
      {smallText && <div className={styles.smallText}>{smallText}</div>}
    </div>
  );
};

export default ResultsInfoContainer;
