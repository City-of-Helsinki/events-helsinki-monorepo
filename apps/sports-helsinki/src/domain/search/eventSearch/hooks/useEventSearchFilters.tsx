import type {
  EventTypeId,
  QueryEventListArgs,
} from 'events-helsinki-components';
import { useRouter } from 'next/router';
import React from 'react';
import { transformedSearchVariables } from '../../combinedSearch/utils';
import { EVENT_SORT_OPTIONS, PAGE_SIZE } from '../constants';
import { getEventSearchVariables } from '../utils';

function useEventSearchFilters(eventType: EventTypeId) {
  const router = useRouter();
  const params: { place?: string; eventType?: string } = router.query;
  return React.useMemo(() => {
    const searchParams = new URLSearchParams(router.asPath.split('?')[1]);
    const transformedParams = transformedSearchVariables(searchParams);
    const variables: QueryEventListArgs = getEventSearchVariables({
      include: ['keywords', 'location'],
      pageSize: PAGE_SIZE,
      params: transformedParams,
      place: params.place,
      sortOrder: EVENT_SORT_OPTIONS.END_TIME,
      superEventType: ['umbrella', 'none'],
      eventType: [eventType],
    });
    return variables;
  }, [router.asPath, params.place, eventType]);
}

export default useEventSearchFilters;
