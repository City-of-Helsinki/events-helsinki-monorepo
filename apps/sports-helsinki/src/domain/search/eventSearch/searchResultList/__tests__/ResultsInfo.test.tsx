import type { AppLanguage } from 'events-helsinki-components';
import * as useLocale from 'events-helsinki-components/hooks/useLocale';

import * as React from 'react';

import { render, screen, userEvent, waitFor } from '@/test-utils';
import ResultsInfo from '../ResultsInfo';

it('events with 0 results matches snapshot for no results', () => {
  const { container } = render(<ResultsInfo resultsCount={0} />);

  expect(container).toMatchSnapshot();
});

it('renders no events found text', async () => {
  render(<ResultsInfo resultsCount={0} />);

  expect(
    screen.getByText(
      'Valitsemillasi hakuehdoilla ei löytynyt yhtään harrastusta'
    )
  ).toBeInTheDocument();
});

it.each([1, 4])(
  'renders few events found text when event count is %i',
  async (resultsCount) => {
    render(<ResultsInfo resultsCount={resultsCount} />);

    const texts = [
      'Valitsemillasi hakuehdoilla löytyi vain vähän hakutuloksia.',
    ];

    texts.forEach((text) => {
      expect(screen.getByText(text)).toBeInTheDocument();
    });
  }
);

it.each<[AppLanguage, number]>([
  ['en', 4],
  ['en', 0],
  ['sv', 0],
])(
  'renders language change button under search results when current language is %s and there are %i %s search items',
  async (language, resultsCount) => {
    // @ts-ignore
    jest.spyOn(useLocale, 'default').mockReturnValue(language);

    const { router } = render(<ResultsInfo resultsCount={resultsCount} />);

    await userEvent.click(
      screen.getByRole('button', { name: 'Näytä hakutulokset suomeksi' })
    );

    await waitFor(() => {
      expect(router.pathname).toBe(`/kurssihaku`);
    });
  }
);

// eslint-disable-next-line max-len
it('renders does not render language change button under eventss search results when current language is Finnish', () => {
  // @ts-ignore
  jest.spyOn(useLocale, 'default').mockReturnValue('fi');

  render(<ResultsInfo resultsCount={0} />);

  expect(
    screen.queryByRole('button', { name: 'Näytä hakutulokset suomeksi' })
  ).not.toBeInTheDocument();
});
