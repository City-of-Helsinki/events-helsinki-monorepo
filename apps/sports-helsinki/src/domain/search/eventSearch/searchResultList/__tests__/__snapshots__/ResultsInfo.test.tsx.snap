// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`events with 0 results matches snapshot for no results 1`] = `
<div>
  <div
    class="_noResultsInfo_d60cd5"
  >
    <div
      class="_iconWrapper_d60cd5"
    >
      <svg
        aria-hidden="true"
        aria-label="search"
        class="Icon-module_icon__1Jtzj icon_hds-icon__1YqNC Icon-module_s__2WGWe icon_hds-icon--size-s__2Lkik"
        role="img"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M15 1C19.4183 1 23 4.58172 23 9C23 13.4183 19.4183 17 15 17C13.2001 17 11.539 16.4056 10.2023 15.4023L3.80101 21.8021L2.21002 20.2111L8.60822 13.8117C7.59861 12.4726 7.00002 10.8062 7.00002 9C7.00002 4.58172 10.5817 1 15 1ZM15 3C11.6863 3 9.00002 5.68629 9.00002 9C9.00002 12.3137 11.6863 15 15 15C18.3137 15 21 12.3137 21 9C21 5.68629 18.3137 3 15 3Z"
          fill="currentColor"
          fill-rule="evenodd"
        />
      </svg>
    </div>
    <div
      class="_bigText_d60cd5"
    >
      Valitsemillasi hakuehdoilla ei löytynyt yhtään paikkaa.
    </div>
  </div>
</div>
`;
